File: lab07/lab07_example01/scripts/lab07_example01_node.py
#!/usr/bin/env python

import rospy
import rospkg
import rosbag
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint

def youbot_traj():

    rospy.init_node('lab7youbot_traj')

    rospack = rospkg.RosPack()

    path = rospack.get_path('lab07_example01')
    bag = rosbag.Bag(path + '/../lab7data.bag')

    traj_pub = rospy.Publisher('/EffortJointInterface_trajectory_controller/command', JointTrajectory, queue_size=5)

    my_traj = JointTrajectory()


    tfs = 4
    my_traj.header.stamp = rospy.Time.now()
    my_traj.joint_names.append('arm_joint_1')
    my_traj.joint_names.append('arm_joint_2')
    my_traj.joint_names.append('arm_joint_3')
    my_traj.joint_names.append('arm_joint_4')
    my_traj.joint_names.append('arm_joint_5')

    for topic, msg, t in bag.read_messages(topics=['my_joint_states']):
        my_pt = JointTrajectoryPoint()

        if len(msg.position) != 0:
            print len(msg.position)
            for i in range(0, 5):
                my_pt.positions.append(msg.position[i])
                my_pt.velocities.append(msg.velocity[i])

        my_pt.time_from_start.secs = tfs
        tfs = tfs + 4
        my_traj.points.append(my_pt)

    bag.close

    rospy.sleep(5)

    traj_pub.publish(my_traj)


if __name__ == '__main__':
    try:
        youbot_traj()
    except rospy.ROSInterruptException:
        pass

File: lab07/lab07example01/src/lab07example01_node.cpp
#include "ros/ros.h"
#include "rosbag/bag.h"
#include "rosbag/view.h"
#include "trajectory_msgs/JointTrajectory.h"
#include "sensor_msgs/JointState.h"

#include "boost/foreach.hpp"


int main(int argc, char **argv)
{

    ros::init(argc, argv, "lab7youbot_traj");

    rosbag::Bag mybag;

    //Specify the mode (Read/Write)
    //Check the definition of "MY_BAG_PATH" in the CMakeLists.txt
    //mybag.open(MY_BAG_PATH, rosbag::bagmode::Read);
    mybag.open(MY_BAG_A, rosbag::bagmode::Read);

    std::vector<std::string> topics;
    topics.push_back(std::string("my_joint_states"));

    trajectory_msgs::JointTrajectory my_traj;
    trajectory_msgs::JointTrajectoryPoint my_pt;

    ros::NodeHandle nh;

    //Initliaise the publisher
    ros::Publisher traj_pub = nh.advertise<trajectory_msgs::JointTrajectory>("/EffortJointInterface_trajectory_controller/command", 5);

    //Initialise all the necessary variables. It is important that the joint names match with the ones described in the urdf file of the robot.
    my_traj.header.stamp = ros::Time::now();
    my_traj.joint_names.push_back("arm_joint_1");
    my_traj.joint_names.push_back("arm_joint_2");
    my_traj.joint_names.push_back("arm_joint_3");
    my_traj.joint_names.push_back("arm_joint_4");
    my_traj.joint_names.push_back("arm_joint_5");

    my_pt.positions.resize(5);
    my_pt.velocities.resize(5);

    rosbag::View view(mybag, rosbag::TopicQuery(topics));

    int tfs = 4;

    BOOST_FOREACH(rosbag::MessageInstance const m, view)
                {
                    sensor_msgs::JointState::ConstPtr J = m.instantiate<sensor_msgs::JointState>();
                    if (J != NULL)
                    {
                        if (J->position.size() != 0)
                        {
                            my_pt.time_from_start.sec = tfs;
                            for (int i = 0; i < 5;i++)
                            {
                                std::cout << "i" << i << std::endl;
                                my_pt.positions.at(i) = J->position.at(i);
                                std::cout << "J->position.at(i)" << J->position.at(i) << std::endl;
                                std::cout << J->position.at(i) << std::endl;
                                my_pt.velocities.at(i) = J->velocity.at(i);
                                std::cout << "J->velocity.at(i)" << J->velocity.at(i) << std::endl;
                                std::cout << J->velocity.at(i) << std::endl;
                            }
                            my_traj.points.push_back(my_pt);
                            tfs = tfs + 4;
                        }
                    }

                }

    //This sleep varies in different machines. This is to prevent the publisher destroying itself before even publishes one trajectory.
    sleep(5);

    traj_pub.publish(my_traj);
    ros::spinOnce();

    mybag.close();
    return 5;
}
